scene/animation/animation_tree.h

```diff
@@ -419,6 +453,27 @@ class AnimationTree : public Node {
 	void advance(double p_time);
 
 	uint64_t get_last_process_pass() const;
+
+	void set_record_buffer_max_size(uint32_t P_record_buffer_max_size);
+	uint32_t get_record_buffer_max_size() const { return record_buffer_max_size; }
+
+	void set_recording_enabled(bool p_enabled) { recording_enabled = p_enabled; }
+	bool is_recording_enabled() const { return recording_enabled; }
+
+	void set_recording_rate(real_t p_rate);
+	real_t get_recording_rate() const { return recording_rate; }
+
+	int get_record_buffer_size() const { return record_buffer.size(); }
+	void resize_record_buffer(int p_size);
+
+	void clear_records() { record_buffer.clear(); }
+
+	void discard_latest_records(uint32_t p_count);
```

Comment:
```suggestion
	void discard_latest_records(uint32_t p_length);
```
Nit pick.

Reply:
If it is possible to implement the offset of the head of the circular queue, I believe it can be more generic, `discard_records(int p_index, int p_length)` with doing offset and loop pop_front() and offset again.

Reply:
I have not idea about this use case (remove records at the middle of queue).

In my opinion, usees can remove outdate records at the end of  queue by calling `resize_record_buffer()`
and discard latest records for rollbacking and than resimulation by calling `discard_records()`.
I think current design can over all use case.

In addition, the purpose of adding `CircularDeque` is to reducing cost of adding/removing element at the front/back, if insert/erase element at the middle of queue, it will lead to move elements like `Vector`, too.

Reply:
> Nit pick.

I think `p_length` is inaccurate.
Maybe `p_discard_count` or `p_discard_number` is better.

Reply:
I have not seen "count" used anywhere else to specify the length of array elements.

Reply:
I decided to use `p_discard_length`.

Reply:
That sounds redundant.

It would also be ideal to have the function `CircularDeque::discard(int p_length, int p_from_index = 0)` in CircularDeque itself, as this is likely to be needed if CircularDeque is used elsewhere.

Reply:
Because other templates are have not a operation for remove elements at the middle of container at once, I will not implement `CircularDeque::discard(int p_length, int p_from_index = 0)`.
Instead, I will implement `insert()` and `remove_at`.